### DISCUSSION TITLE
Fyne app compiled for Windows expects MS OpenGL driver? So Wine on MacOS cannot run a Fyne app?

### DISCUSSION THREAD

**Discussion by Gys** [fyne-io/fyne#4905] [General] [answered]
I am developing on a MacBook 2018 (x86) running Sonoma 14.5

The app that I made should also run on Windows. To test this easily I tried to use Wine, a Windows emulator also running on MacOS (https://www.winehq.org/). 

If I run the cross compiled .exe with Wine it shows at some point: `panic: failed to link OpenGL program; error code: 506`.

So I understand Fyne uses OpenGL for rendering. I assume for Windows that means using a native Microsoft driver? A driver that Wine not has when running on MacOS?

I just want to make sure Wine is indeed unusable for testing Fyne Windows applications.

**Comment by andydotxyz** [comment, marked as answer]
Don't use an emulator to test for different devices. A virtual machine should work though (just remember to turn on the OpenGL/3D acceleration support).

**Reply by Gys** [reply]
I moved the windows exe to a virtual machine on Azure, running Windows 11. Still an error related to OpenGL:
![Screenshot 2024-06-03 at 16 51 52](https://github.com/fyne-io/fyne/assets/943251/45988f01-f2eb-4ad2-9153-acc274e82c49)
I also compiled my app using fyne-cross, to see if compiling makes a difference. That resulted in a similar error:
![Screenshot 2024-06-03 at 17 32 58](https://github.com/fyne-io/fyne/assets/943251/f2329d5b-6284-4753-acb6-2c52e4c64abe)

It seems I cannot configure anything related to OpenGL in the settings of the Azure VM.

**Reply by Gys** [reply]
I then installed Go on the Windows 11 VM, ran `go mod tidy` and tried to compile the source. That resulting in the message below. Why would this report build constraints? I would assume those are transparent on the platform itself?

````
package guidu
        imports fyne.io/fyne/v2/app
        imports fyne.io/fyne/v2/internal/driver/glfw
        imports fyne.io/fyne/v2/internal/driver/common
        imports fyne.io/fyne/v2/internal/painter/gl
        imports github.com/go-gl/gl/v2.1/gl: build constraints exclude all Go files in C:\Users\gijs\go\pkg\mod\github.com\go-gl\gl@v0.0.0-20211210172815-726fda9656d6\v2.1\gl
````

**Reply by andydotxyz** [reply]
> I moved the windows exe to a virtual machine on Azure

Azure VMs don't have graphics drivers either - it would need to be either a local VM or you would have to upload some software based graphics driver into the cloud.

> That resulting in the message below. Why would this report build constraints? I would assume those are transparent on the platform itself?

This message actually means that there is no C compiler installed. I know it's a little confusing but we can't intercept that message as it's Go saying there is no way to build the C code. Whatever developer setup you had locally would need to translate to the cloud to build it there.

However building on the VM and building locally then uploading will not have any different outcome - the compiled apps are portable.

**Reply by Gys** [reply]
I installed VMWare Fusion and downloaded Windows 10 from Microsoft (48 hours test version?!). Indeed then the app works as expected.

**Comment by andydotxyz** [comment]
The "OpenGL" driver is not Microsoft specific, it is provided by the graphics card.
